Add assumption docs for provisioning tools (#514)
kubernetes-sigs/cloud-provider-azure: site/content/en/topics/assumptions.md

---
title: "Assumptions for Provisioning Tools"
linkTitle: "Assumptions for Provisioning Tools"
weight: 1
type: docs
description: >
    Cloud provider assumptions on Azure resources that provisioning tools should take care.
---

Here is a list of Azure resource assumptions that are required for cloud provider Azure:

* All Azure resources are under the same tenant.
* All the virtual machines are part of either VirtualMachineScaleSet or AvailabilitySet.
* All virtual machine's name should be the same as their hostname.

> * LoadBalancer's name should be `<clusterName>` for external type and `<clusterName>-internal` for internal type.

Reviewer: What about multiple SLB scenario? In the future, the expected way would be:
1. each VMSS or VMAS owns an external SLB;
2. only the SLB of the primary VMSS or VMAS would be named `<clusterName>`, others would be named `<vm_set_name>`.

Author: let me add a set for it.

  * `<clusterName>` is coming from `--cluster-name` configuration (default is `kubernetes`).

After provisioned, cloud provider Azure would touch the following Azure resources based on workloads:

* New routes would be added for each node if `--configure-cloud-routes` is enabled.
* New LoadBalancer (including external and internal) would be created if they're not existing yet.

> * Virtual machines would be added to LoadBalancer backend address pools if they're not added yet.

Reviewer: The LB ref ID would be added to VM nics and VMSS as well.

---

> title: "Assumptions for Provisioning Tools"

Reviewer 2: ```suggestion
title: "Cluster Provisioning Tools Contract"
```

linkTitle: "Assumptions for Provisioning Tools"
weight: 1

> type: docs

Reviewer 2: not sure how this formatting works but is it possible to add a "last modified" or "last revised" date in here?

Author: last modified date is generated from commit, hence it is not needed here.

description: >

>     Cloud provider assumptions on Azure resources that provisioning tools should take care.

Reviewer 2: ```suggestion
    Cloud provider assumptions on Azure resources that provisioning tools should follow.
```

---

> 

Reviewer 2: ```suggestion
0The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
```

Here is a list of Azure resource assumptions that are required for cloud provider Azure:

> * All Azure resources are under the same tenant.

Reviewer 2: let's follow RFC 2119 here https://tools.ietf.org/html/rfc2119 

eg. `All Azure resources MUST be under the same tenant.`

> * All the virtual machines are part of either VirtualMachineScaleSet (VMSS) or AvailabilitySet (VMAS).

Reviewer 2: This we need to rework. There is no Azure requirement for Virtual Machines to be in an availability set or scale set. Availability sets are one of the options for HA but not the only one. In regions that support them, spreading VMs on availability zones is a better option to ensure availability. Availability zones is a separate physical zone with its own power source, network, and cooling. On the other hand, an availability set only provides different fault domains, which are logical hardware groups. In the event that an entire datacenter building is affected, a zone provides better resiliency, if available.

https://docs.microsoft.com/en-us/azure/virtual-machines/availability 

In fact, CAPZ already does this today by spreading virtual machines on zones if available, only putting them in availability set if zones are not available. It works wit cloud provider today so we need to ensure it doesn't break in the future.

Reviewer 2: related to https://github.com/kubernetes/kubernetes/issues/97375

Author: VMAS/VMSS are not conflicted with AZ, they are different levels of ensuring the HA. VMAS/VMSS are required because some of cloud provider functionality need concepts of "node pool"

Reviewer 2: Availability sets and Availability zones are two options for availability, you wouldn't generally use both, it's one or the other. Some k8s provisioning tools like cluster api (or openshift) have different ways to define a "node pool" that are not synonymous with an availability set or a scale set, like a machine deployment. In CAPZ, a machine deployment will result in VMs being spread in all available zones and not placed in an availability set. A standalone VM without AS should be a supported use case (and it already is today).

cc @wking who commented on this in kubernetes/kubernetes#97375

Author: makes sense, thanks for the explanation. Moved this under enableMultipleStandardLoadBalancers.

> * All virtual machine's name should be the same as their hostname.

Reviewer 2: does it apply for Windows? cc @jsturtevant

Author: yes

> * LoadBalancer's names should be following rules (`<clusterName>` is coming from `--cluster-name` configuration, default is `kubernetes`)

Reviewer 2: can we specify which LB we're talking about here? Provisioning tools often create more than one LB (eg. one for the apiserver, one for node outbound - which also serves as the service LB used by cloud provider)

Author: this is for node LoadBalancer

  * When `enableMultipleStandardLoadBalancers` is configured to `false`, LoadBalancer's name should be `<clusterName>` for external type and `<clusterName>-internal` for internal type.
  * When `enableMultipleStandardLoadBalancers` is configured to `true`, multiple standard load balancers should be provisioned:
    * Each VMAS and VMSS should be put behind a different standard LoadBalancer.
    * The primary LoadBalancer's name should be `<clusterName>` for external type and `<clusterName>-internal` for internal type. Virtual machines that are part of primary VMAS (set by `primaryAvailabilitySetName`) or primary VMSS (set by `primaryScaleSetName`) would be added to primary LoadBalancer backend address pool.
    * Other standard LoadBalancer's name should be same as VMAS and VMSS name.

> After provisioned, cloud provider Azure would touch the following Azure resources based on workloads:

Reviewer 2: ```suggestion
After the cluster is provisioned, cloud provider Azure CAN update the following Azure resources based on workloads:
```

---
title: "Cluster Provisioning Tools Contract"
linkTitle: "Cluster Provisioning Tools Contract"
weight: 1
type: docs
description: >
    Cloud provider assumptions on Azure resources that provisioning tools should follow.
---

> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

Here is a list of Azure resource assumptions that are required for cloud provider Azure:

* All Azure resources MUST be under the same tenant.
* All the virtual machines MUST be part of either VirtualMachineScaleSet (VMSS) or AvailabilitySet (VMAS).

> * All virtual machine's name should be the same as their hostname.

Reviewer 2: ```suggestion
* All virtual machine names MUST be the same as their hostname.
```

  * When `enableMultipleStandardLoadBalancers` is configured to `true`, multiple standard load balancers SHOULD be provisioned:
    * All the virtual machines MUST be part of either VirtualMachineScaleSet (VMSS) or AvailabilitySet (VMAS).
    * Each VMAS and VMSS SHOULD be put behind a different standard LoadBalancer.
    * The primary LoadBalancer's name SHOULD be `<clusterName>` for external type and `<clusterName>-internal` for internal type. Virtual machines that are part of primary VMAS (set by `primaryAvailabilitySetName`) or primary VMSS (set by `primaryScaleSetName`) SHOULD be added to primary LoadBalancer backend address pool.
    * Other standard LoadBalancer's name SHOULD be same as VMAS or VMSS name.

After the cluster is provisioned, cloud provider Azure MAY update the following Azure resources based on workloads:

* New routes would be added for each node if `--configure-cloud-routes` is enabled.
* New LoadBalancer (including external and internal) would be created if they're not existing yet.

> * Virtual machines and virtual machine scale sets would be added to LoadBalancer backend address pools if they're not added yet.

Reviewer 2: does it make sense to also add backend pool name somewhere in here since we assume <cluster-name> for ipv4 and <cluster-name-IPv6> for IPv6?

Author: provisioning tools could use different backend pool from cloud provider (e.g. AKS has created different backend pool for outbound rules).

Reviewer 2: right but you don't want to name it the same thing if you intend them to be separate, otherwise, cloud provider will reuse it and modify it

Author: right, so it is not a required config. The name should depend on the detailed scenarios.
